File: apps/conclave-skip/Sources/Conclave/Core/Layout/GridLayout.swift
//
//  GridLayout.swift
//  Conclave
//
//  Google-Meet-style participant grid arrangement — a faithful 1:1 port of the
//  web/RN engine in `packages/meeting-core/src/grid-layout.ts`. Pure, no UI deps,
//  so iOS (SwiftUI) and Android (Compose via Skip) lay tiles out IDENTICALLY to
//  the web app. Given a tile count, a measured container, and a target aspect,
//  it finds the column/row arrangement that MAXIMISES tile size (the canonical
//  optimal-grid packing Meet/Jitsi/Daily use), then reports last-row centering
//  and paging.
//
//  Keep this in lockstep with grid-layout.ts — same objective, same tie-breaks.
//

import Foundation
#if canImport(CoreGraphics)
import CoreGraphics
#endif

struct GridLayoutOptions {
    /// Gap between tiles in pt. Subtracted from the container before dividing.
    var gap: CGFloat = 12
    /// Hard cap on columns (device class: ~3 phone portrait, 4–6 tablet/landscape).
    var maxCols: Int = 7
    /// Tiles shown per page before paging/overflow kicks in.
    var maxTilesPerPage: Int = 49
    /// Target tile aspect ratio = width / height (16:9 = 1.7778 landscape).
    var targetAspect: CGFloat = 16.0 / 9.0
}

struct GridTilePosition {
    /// Zero-based tile order on the current page.
    let index: Int
    /// Zero-based visual row.
    let row: Int
    /// Zero-based visual column within the logical grid. Partial rows may be fractional.
    let col: CGFloat
    /// Left offset inside the measured layout area, in pt.
    let x: CGFloat
    /// Top offset inside the measured layout area, in pt.
    let y: CGFloat
    /// Tile width in pt.
    let width: CGFloat
    /// Tile height in pt.
    let height: CGFloat
}

struct GridLayoutResult {
    /// Columns and rows for a full page.
    let cols: Int
    let rows: Int
    /// Size of each tile in pt (already aspect-constrained to targetAspect).
    let tileWidth: CGFloat
    let tileHeight: CGFloat
    /// Tiles in the final (possibly partial) row — render centered.
    let lastRowCount: Int
    /// Number of pages when count exceeds maxTilesPerPage.
    let pages: Int
    /// Tiles laid out on a full page (<= maxTilesPerPage).
    let perPage: Int
    /// Width of the centered tile group, in pt.
    let contentWidth: CGFloat
    /// Height of the centered tile group, in pt.
    let contentHeight: CGFloat
    /// Left offset of the centered tile group inside the measured layout area.
    let offsetX: CGFloat
    /// Top offset of the centered tile group inside the measured layout area.
    let offsetY: CGFloat
    /// Exact tile positions for the current page, including centered partial rows.
    let positions: [GridTilePosition]
}

private struct GridCandidate {
    let cols: Int
    let rows: Int
    let tileWidth: CGFloat
    let empty: Int
    let aspectDist: CGFloat
}

private func buildGridTilePositions(
    cols: Int,
    rows: Int,
    perPage: Int,
    tileWidth: CGFloat,
    tileHeight: CGFloat,
    gap: CGFloat,
    width: CGFloat,
    height: CGFloat
) -> (
    contentWidth: CGFloat,
    contentHeight: CGFloat,
    offsetX: CGFloat,
    offsetY: CGFloat,
    positions: [GridTilePosition]
) {
    let contentWidth = max(0.0, CGFloat(cols) * tileWidth + CGFloat(max(0, cols - 1)) * gap)
    let contentHeight = max(0.0, CGFloat(rows) * tileHeight + CGFloat(max(0, rows - 1)) * gap)
    let offsetX = max(0.0, (width - contentWidth) / 2.0)
    let offsetY = max(0.0, (height - contentHeight) / 2.0)
    var positions: [GridTilePosition] = []

    var index = 0
    while index < perPage {
        let row = index / cols
        let rowStartIndex = row * cols
        let rowCount = min(cols, perPage - rowStartIndex)
        let rowWidth = CGFloat(rowCount) * tileWidth + CGFloat(max(0, rowCount - 1)) * gap
        let rowOffsetX = offsetX + max(0.0, (contentWidth - rowWidth) / 2.0)
        let col = index - rowStartIndex

        positions.append(GridTilePosition(
            index: index,
            row: row,
            col: CGFloat(col) + CGFloat(max(0, cols - rowCount)) / 2.0,
            x: rowOffsetX + CGFloat(col) * (tileWidth + gap),
            y: offsetY + CGFloat(row) * (tileHeight + gap),
            width: tileWidth,
            height: tileHeight
        ))
        index += 1
    }

    return (contentWidth, contentHeight, offsetX, offsetY, positions)
}

/// Find the arrangement of `count` aspect-locked tiles that fits the largest
/// tile inside a `width` × `height` container. Objective: maximise the displayed
/// video area; ties break toward the grid whose bounding box best matches the
/// container aspect (so 2 people sit side-by-side in landscape), then fewer
/// empty cells.
func computeGridLayout(
    count: Int,
    width: CGFloat,
    height: CGFloat,
    options: GridLayoutOptions = GridLayoutOptions()
) -> GridLayoutResult {
    let gap = options.gap
    let targetAspect = options.targetAspect

    let total = max(1, count)
    let pages = Int(ceil(Double(total) / Double(options.maxTilesPerPage)))
    let perPage = min(total, options.maxTilesPerPage)

    // Degenerate container — return a single column so the caller still renders.
    if !width.isFinite || !height.isFinite || width <= 0 || height <= 0 {
        let placement = buildGridTilePositions(
            cols: 1, rows: perPage, perPage: perPage,
            tileWidth: 0.0, tileHeight: 0.0, gap: gap, width: 0.0, height: 0.0
        )
        return GridLayoutResult(
            cols: 1, rows: perPage, tileWidth: 0.0, tileHeight: 0.0,
            lastRowCount: 1, pages: pages, perPage: perPage,
            contentWidth: placement.contentWidth,
            contentHeight: placement.contentHeight,
            offsetX: placement.offsetX,
            offsetY: placement.offsetY,
            positions: placement.positions
        )
    }

    let colCap = min(perPage, max(1, options.maxCols))
    let containerAspect = width / height

    var best: GridCandidate? = nil

    var cols = 1
    while cols <= colCap {
        let rows = Int(ceil(Double(perPage) / Double(cols)))
        let cellW = (width - CGFloat(cols - 1) * gap) / CGFloat(cols)
        let cellH = (height - CGFloat(rows - 1) * gap) / CGFloat(rows)
        if cellW > 0 && cellH > 0 {
            // Largest aspect-locked tile that fits the cell (letterbox-fit).
            let tileWidth = min(cellW, cellH * targetAspect)
            let tileHeight = tileWidth / targetAspect
            let empty = cols * rows - perPage
            let boxW = CGFloat(cols) * tileWidth + CGFloat(cols - 1) * gap
            let boxH = CGFloat(rows) * tileHeight + CGFloat(rows - 1) * gap
            let aspectDist = abs(boxW / boxH - containerAspect)

            var better = best == nil
            if let b = best {
                if tileWidth > b.tileWidth + 0.5 {
                    better = true
                } else if abs(tileWidth - b.tileWidth) <= 0.5 {
                    if aspectDist < b.aspectDist - 0.01 {
                        better = true
                    } else if abs(aspectDist - b.aspectDist) <= 0.01 && empty < b.empty {
                        better = true
                    }
                }
            }
            if better {
                best = GridCandidate(cols: cols, rows: rows, tileWidth: tileWidth, empty: empty, aspectDist: aspectDist)
            }
        }
        cols += 1
    }

    let chosen = best ?? GridCandidate(cols: 1, rows: perPage, tileWidth: width, empty: 0, aspectDist: 0.0)
    let tileWidth = max(0.0, floor(chosen.tileWidth))
    let tileHeight = max(0.0, floor(tileWidth / targetAspect))
    let lastRowCount = perPage - (chosen.rows - 1) * chosen.cols
    let placement = buildGridTilePositions(
        cols: chosen.cols, rows: chosen.rows, perPage: perPage,
        tileWidth: tileWidth, tileHeight: tileHeight, gap: gap, width: width, height: height
    )

    return GridLayoutResult(
        cols: chosen.cols,
        rows: chosen.rows,
        tileWidth: tileWidth,
        tileHeight: tileHeight,
        lastRowCount: max(1, lastRowCount),
        pages: pages,
        perPage: perPage,
        contentWidth: placement.contentWidth,
        contentHeight: placement.contentHeight,
        offsetX: placement.offsetX,
        offsetY: placement.offsetY,
        positions: placement.positions
    )
}

// MARK: - Stage mode (Meet "Auto" selection above the grid packer)

enum StageMode {
    case tiled
    case spotlight
    case sideBySide
    case sidebar
}

struct StageModeInput {
    /// Visible participant count (incl. self).
    let count: Int
    /// Someone is screen-sharing / presenting.
    let presenting: Bool
    /// A tile is explicitly pinned to the stage.
    let pinned: Bool
    /// There is an active *video* speaker (not just the presentation).
    let hasActiveVideoSpeaker: Bool
    /// Above this count (no presentation) Meet uses a sidebar instead of tiled.
    var tiledThreshold: Int = 12
}

/// Meet "Auto" mode selection that runs ABOVE the grid packer. The packer is the
/// engine for `tiled` and for the people rail in `sidebar`/`sideBySide`.
func chooseStageMode(_ input: StageModeInput) -> StageMode {
    if input.pinned || input.count <= 2 || (input.presenting && !input.hasActiveVideoSpeaker) {
        return .spotlight
    }
    if input.presenting {
        return .sideBySide
    }
    if input.count <= input.tiledThreshold {
        return .tiled
    }
    return .sidebar
}

File: packages/meeting-core/src/grid-layout.ts
/**
 * Google-Meet-style participant grid arrangement — shared, pure, dependency-free.
 *
 * This is the client-side "tiled / auto" layout engine: given a visible tile
 * count, a measured container, and a target tile aspect ratio, it finds the
 * column/row arrangement that MAXIMISES tile size (the canonical optimal-grid
 * packing that Meet/Jitsi/Daily all use), then reports last-row centering and
 * paging so the renderer can lay it out exactly like Meet.
 *
 * One source of truth: web (apps/web) and React-Native (apps/mobile) import
 * this; the Skip/SwiftUI app mirrors it 1:1 in Swift (GridLayout.swift).
 */

export interface GridLayoutOptions {
  /** Gap between tiles in px. Subtracted from the container before dividing. */
  gap?: number;
  /** Hard cap on columns (device class: ~6–7 desktop, 2–3 narrow phone). */
  maxCols?: number;
  /** Tiles shown per page before paging/overflow kicks in. */
  maxTilesPerPage?: number;
  /** Target tile aspect ratio = width / height (16:9 = 1.7778 landscape). */
  targetAspect?: number;
}

export interface GridLayoutResult {
  /** Columns and rows for a full page. */
  cols: number;
  rows: number;
  /** Size of each tile in px (already aspect-constrained to targetAspect). */
  tileWidth: number;
  tileHeight: number;
  /** Tiles in the final (possibly partial) row — render centered. */
  lastRowCount: number;
  /** Number of pages when count exceeds maxTilesPerPage. */
  pages: number;
  /** Tiles laid out on a full page (<= maxTilesPerPage). */
  perPage: number;
}

const DEFAULTS: Required<GridLayoutOptions> = {
  gap: 12,
  maxCols: 7,
  maxTilesPerPage: 49,
  targetAspect: 16 / 9,
};

/**
 * Find the arrangement of `count` aspect-locked tiles that fits the largest
 * tile inside a `width` × `height` container.
 *
 * The objective is "maximise the side of the largest tile that fits N copies"
 * (i.e. maximise the displayed-video area). For each candidate column count we
 * derive rows = ceil(N/cols), compute the largest tile that fits the cell while
 * respecting `targetAspect`, and keep the candidate whose tile is biggest;
 * ties break toward fewer empty cells (fuller grid).
 */
export function computeGridLayout(
  count: number,
  width: number,
  height: number,
  options: GridLayoutOptions = {},
): GridLayoutResult {
  const { gap, maxCols, maxTilesPerPage, targetAspect } = { ...DEFAULTS, ...options };

  const total = Math.max(1, Math.floor(count));
  const pages = Math.ceil(total / maxTilesPerPage);
  const perPage = Math.min(total, maxTilesPerPage);

  // Degenerate container — return a single column so the caller still renders.
  if (!Number.isFinite(width) || !Number.isFinite(height) || width <= 0 || height <= 0) {
    return { cols: 1, rows: perPage, tileWidth: 0, tileHeight: 0, lastRowCount: 1, pages, perPage };
  }

  const colCap = Math.min(perPage, Math.max(1, Math.floor(maxCols)));
  const containerAspect = width / height;

  let best:
    | { cols: number; rows: number; tileWidth: number; empty: number; aspectDist: number }
    | null = null;

  for (let cols = 1; cols <= colCap; cols++) {
    const rows = Math.ceil(perPage / cols);
    const cellW = (width - (cols - 1) * gap) / cols;
    const cellH = (height - (rows - 1) * gap) / rows;
    if (cellW <= 0 || cellH <= 0) continue;

    // Largest aspect-locked tile that fits the cell (letterbox-fit).
    const tileWidth = Math.min(cellW, cellH * targetAspect);
    const tileHeight = tileWidth / targetAspect;
    const empty = cols * rows - perPage;
    // How well the grid's bounding box matches the container — this is the
    // tie-break that makes e.g. 2 people sit side-by-side in a landscape
    // container (both arrangements give the same tile size; the wider box wins).
    const boxW = cols * tileWidth + (cols - 1) * gap;
    const boxH = rows * tileHeight + (rows - 1) * gap;
    const aspectDist = Math.abs(boxW / boxH - containerAspect);

    let better = best === null;
    if (best !== null) {
      if (tileWidth > best.tileWidth + 0.5) better = true;
      else if (Math.abs(tileWidth - best.tileWidth) <= 0.5) {
        // Near-tie on tile size → best aspect match, then fuller grid.
        if (aspectDist < best.aspectDist - 0.01) better = true;
        else if (Math.abs(aspectDist - best.aspectDist) <= 0.01 && empty < best.empty)
          better = true;
      }
    }
    if (better) best = { cols, rows, tileWidth, empty, aspectDist };
  }

  // Fallback (shouldn't happen): one column.
  const chosen = best ?? { cols: 1, rows: perPage, tileWidth: width };

  const tileWidth = Math.max(0, Math.floor(chosen.tileWidth));
  const tileHeight = Math.max(0, Math.floor(tileWidth / targetAspect));
  const lastRowCount = perPage - (chosen.rows - 1) * chosen.cols;

  return {
    cols: chosen.cols,
    rows: chosen.rows,
    tileWidth,
    tileHeight,
    lastRowCount: Math.max(1, lastRowCount),
    pages,
    perPage,
  };
}

export type StageMode = "tiled" | "spotlight" | "sideBySide" | "sidebar";

export interface StageModeInput {
  /** Visible participant count (incl. self). */
  count: number;
  /** Someone is screen-sharing / presenting. */
  presenting: boolean;
  /** A tile is explicitly pinned to the stage. */
  pinned: boolean;
  /** There is an active *video* speaker (not just the presentation). */
  hasActiveVideoSpeaker: boolean;
  /** Above this count (no presentation) Meet uses a sidebar instead of tiled. */
  tiledThreshold?: number;
}

/**
 * Meet "Auto" mode selection that runs ABOVE the grid packer. The packer is the
 * engine for the `tiled` mode and for the people rail in `sidebar`/`sideBySide`.
 */
export function chooseStageMode(input: StageModeInput): StageMode {
  const { count, presenting, pinned, hasActiveVideoSpeaker, tiledThreshold = 12 } = input;
  if (pinned || count <= 2 || (presenting && !hasActiveVideoSpeaker)) return "spotlight";
  if (presenting) return "sideBySide";
  if (count <= tiledThreshold) return "tiled";
  return "sidebar";
}
